src/Pate/Verification/AbstractDomain.hs

```diff
@@ -0,0 +1,506 @@
+        unless (absValCombined == absValPrev) $ CMW.tell (WidenLocs (S.singleton (Some reg)) mempty)
+        return absValCombined
+
+widenAbsDomainVals ::
+  forall sym arch m.
+  Monad m =>
+  IO.MonadIO m =>
+  W4.IsSymExprBuilder sym =>
+  PA.ValidArch arch =>
+  sym ->
+  AbstractDomainBody sym arch {- ^ existing abstract domain that this is augmenting -} ->
+  (forall tp. W4.SymExpr sym tp -> m (AbsRange tp)) ->
+  PS.SimBundle sym arch ->
+  m (AbstractDomainBody sym arch, Maybe (WidenLocs sym arch))
+widenAbsDomainVals sym prev f bundle = do
+  let (PPa.PatchPair absValsO absValsP) = absDomVals prev
+  (absValsO', locsO) <- widenAbsDomainVals' sym absValsO f (PS.simOutO bundle)
+  (absValsP', locsP) <- widenAbsDomainVals' sym absValsP f (PS.simOutP bundle)
+  return $ (prev { absDomVals = PPa.PatchPair absValsO' absValsP' }, Just $ locsO <> locsP)
+
+
+applyAbsRange ::
+  W4.IsSymExprBuilder sym =>
+  sym ->
+  W4.SymExpr sym tp ->
+  AbsRange tp ->
+  IO (PS.AssumptionFrame sym)
+applyAbsRange sym e rng = case rng of
+  AbsIntConstant i -> do
+    i' <- W4.intLit sym i
+    return $ PS.exprBinding e i'
+  AbsBVConstant w bv -> do
+    bv' <- W4.bvLit sym w bv
+    return $ PS.exprBinding e bv'
+  AbsBoolConstant True -> return $ PS.exprBinding e (W4.truePred sym)
+  AbsBoolConstant False -> return $ PS.exprBinding e (W4.falsePred sym)
+  AbsUnconstrained{} -> return mempty
+
+absDomainValToAsm ::
+  W4.IsSymExprBuilder sym =>
+  sym ->
+  PSR.MacawRegEntry sym tp ->
+  MacawAbstractValue sym tp ->
+  IO (PS.AssumptionFrame sym)
+absDomainValToAsm sym e (MacawAbstractValue absVal) = case PSR.macawRegRepr e of
+  CLM.LLVMPointerRepr{} -> do
+    CLM.LLVMPointer region offset <- return $ PSR.macawRegValue e
+    (Ctx.Empty Ctx.:> regAbs Ctx.:> offsetAbs) <- return $ absVal
+    regionInt <- W4.natToInteger sym region
+    regFrame <- applyAbsRange sym regionInt regAbs
+    offFrame <- applyAbsRange sym offset offsetAbs
+    return $ regFrame <> offFrame
+  CT.BoolRepr -> do
+    b <- return $ PSR.macawRegValue e
+    (Ctx.Empty Ctx.:> bAbs) <- return $ absVal
+    applyAbsRange sym b bAbs
+  CT.StructRepr Ctx.Empty -> return mempty
+  _ -> error "applyAbsDomainVal: unexpected type for abstract domain"
+
+
+-- | Construct an 'PS.AssumptionFrame' asserting
+-- that the values in the given initial state
+-- ('PS.SimInput') are necessarily in the given abstract domain
+absDomainValsToAsm ::
+  forall sym arch bin.
+  W4.IsSymExprBuilder sym =>
+  MapF.OrdF (W4.SymExpr sym) =>
+  MC.RegisterInfo (MC.ArchReg arch) =>
+  sym ->
+  PS.SimState sym arch bin ->
+  AbstractDomainVals sym arch bin ->
+  IO (PS.AssumptionFrame sym)
+absDomainValsToAsm sym st vals = do
+  memFrame <- MapF.foldrMWithKey getCell mempty (absMemVals vals)
+  regFrame <- fmap PRt.collapse $ PRt.zipWithRegStatesM (PS.simRegs st) (absRegVals vals) $ \_ val absVal ->
+    Const <$> absDomainValToAsm sym val absVal
+  return $ memFrame <> regFrame
+  where
+    getCell ::
```

Comment:
It might be nice to call this something like `accumulateCell` given that it is building up the result in a fold